xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp

```diff
@@ -123,14 +132,67 @@ bool CVAAPIContext::EnsureContext(CVAAPIContext **ctx, CDecoder *decoder)
   return true;
 }
 
-bool CVAAPIContext::CreateContext()
+void CVAAPIContext::SetValidDRMVaDisplayFromRenderNode()
+{
+  int const buf_size{128};
+  char name[buf_size];
+  int fd{-1};
+
+  // 128 is the start of the NUM in renderD<NUM>
+  for (int i = 128; i < (128 + 16); i++)
+  {
+    snprintf(name, buf_size, "/dev/dri/renderD%u", i);
+
+    fd = open(name, O_RDWR);
+
+    if (fd < 0)
+    {
+      continue;
+    }
+
+    auto display = vaGetDisplayDRM(fd);
+
+    if (display != nullptr)
+    {
+      m_render_node_fd = fd;
+      m_display = display;
+      return;
+    }
+  }
+
+  CLog::Log(LOGERROR, "FAILED To find any open render nodes in /dev/dri/renderD<num>\n");
+}
+
+void CVAAPIContext::SetVaDisplayForSystem()
 {
-  { CSingleLock lock(g_graphicsContext);
-    if (!m_X11dpy)
-      m_X11dpy = XOpenDisplay(NULL);
+  auto win_system = g_Windowing.GetWinSystem();
+  if (win_system == WINDOW_SYSTEM_X11)
+  {
+#if HAVE_X11
+    Display *x11_display{nullptr};
+    { CSingleLock lock(g_graphicsContext);
+      x11_display = XOpenDisplay(nullptr);
```

Comment:
this opens a display connection every time we start some video. this is not what we want here. we want open it once for the lifetime of the application.